--- a/kernel/kernel_umount.c	2025-11-12 13:58:41.449830789 +0800
+++ b/kernel/kernel_umount.c	2025-11-13 04:24:09.993972336 +0800
@@ -134,6 +134,31 @@
     if (!ksu_kernel_umount_enabled) {
         return 0;
     }
+ 
+#ifdef CONFIG_KSU_SUSFS
+    bool is_zygote_child = susfs_is_sid_equal(old->security, susfs_zygote_sid);
+    if (is_some_system_uid(new_uid.val) && is_zygote_child) {
+        if (ksu_is_allow_uid(new_uid.val)) {
+            return 0;
+        }
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+        susfs_set_current_proc_su_not_allowed();
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_SU
+        if (susfs_is_umount_for_zygote_system_process_enabled) {
+#ifdef CONFIG_KSU_SUSFS
+        susfs_set_current_non_root_user_app_proc();
+#endif // #ifdef CONFIG_KSU_SUSFS
+#ifdef CONFIG_KSU_SUSFS_SUS_SU
+        susfs_set_current_proc_su_not_allowed();
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_SU
+#ifdef CONFIG_KSU_SUSFS_SUS_PATH
+        susfs_run_sus_path_loop(new_uid.val);
+#endif // #ifdef CONFIG_KSU_SUSFS_SUS_PATH
+            goto do_umount;
+        }
+    }
+#endif // #ifdef CONFIG_KSU_SUSFS
+
 
     // FIXME: isolated process which directly forks from zygote is not handled
     if (!is_appuid(new_uid)) {
@@ -144,10 +169,15 @@
         return 0;
     }
 
+do_umount:
     // check old process's selinux context, if it is not zygote, ignore it!
     // because some su apps may setuid to untrusted_app but they are in global mount namespace
     // when we umount for such process, that is a disaster!
+#ifdef CONFIG_KSU_SUSFS
+    // bool is_zygote_child = is_zygote_child;  // provided externally in SUSFS
+#else
     bool is_zygote_child = is_zygote(get_current_cred());
+#endif
     if (!is_zygote_child) {
         pr_info("handle umount ignore non zygote child: %d\n", current->pid);
         return 0;
